### PR TITLE
Add a lint `unconditional_recursion` to detect unconditional recursion.

### LINKED ISSUE #17899: Add lint for direct recursive self-call of function
With trait objects it is not so rare to have code like

```rust
impl Foo for Box<Foo> {
    fn bar(&self, x: uint) {
        // call the `bar` method in the vtable, could try:
        // self.bar(x);
        // (*self).bar(x);
        // (**self).bar(x);
    }
}
```

One might first write `self.bar(x)`, but this will be calling that same bar, via since `Box<Foo>` has a "real" `bar` method itself. The correct procedure is `(**self).bar(x)`. This would helped by detecting the most obvious cases of a function calling itself in a nonterminating way. The most obvious way that is likely to have low risk of false positives is detecting if the CFG of a function reaches a self-call unconditionally. e.g.

```rust
// these receive a warning:
fn foo() { foo() }

fn bar(x: uint) {
    bar(x - 1);
}

fn baz(x: uint) -> uint {
    if x == 0 {
        println!("hi")
    }
    2 * baz(x + 1)
}

// this is fine:
fn qux(x: uint) {
    if x == 0 { loop {} } // (could be `return 1`, etc. etc.)

    qux(x - 1) + 1
}
```

Notably, it does not matter if the arguments to the self call are different, since the call will always occur, and it also does not matter if the self-call is in tail position.

There is the possibility that someone is using infinite recursion to achieve an infinite loop, it doesn't seem crazy to recommend that they use `loop` instead.

A more conservative alternative would be just detecting the method call in the first example, that is, a function that consists of a single expression that is a self-call.

### PR DESCRIPTION
**Pull request by huonw**
E.g. `fn foo() { foo() }`, or, more subtlely

```
impl Foo for Box<Foo+'static> {
    fn bar(&self) {
        self.bar();
    }
}
```

The compiler will warn and point out the points where recursion occurs,
if it determines that the function cannot return without calling itself.

Closes #17899.

### CODE REVIEW

**Comment by huonw** [comment]
@luqmana, speaking prescriptively, "recurs" is actually *more* correct than recurses, e.g. http://english.stackexchange.com/questions/163446/does-a-recursive-procedure-recur ; I'm happy to change it if people prefer the recurse version.

**Comment by huonw** [comment]
@frewsxcv (Thanks! I'm actually leaving it unrebased so it can't be r+'d and landed before I resolve the questions described above.)

**Comment by huonw** [comment]
Updated after discussions with @nikomatsakis.

(Also, this found a bug in the compiler. :smile:)

I measured the total time required to construct all the CFGs in each crate (sorted from slowest to fastest), I think it's safe to say that I do not need to worry about the performance of this at the moment since even constructing the graphs for all functions in librustc is only 50 milliseconds.

| crate |    milliseconds |
|-------|-----------------|
| librustc |   51.3 |
| libsyntax |   40.1 |
| librustc_trans |   29.9 |
| libcore |   15.9 |
| librustc_typeck |   14.1 |
| libstd |   12.6 |
| librustdoc |   11.7 |
| librustc_resolve |   6.8 |
| libserialize |   5.4 |
| libcollections |   4.9 |
| librustc_borrowck |   3.6 |
| librustc_back |   2.5 |
| libregex |   2.4 |
| libtest |   1.5 |
| libgetopts |   1.4 |
| librustc_driver |   1.3 |
| librbml |   1.3 |
| libterm |   1.3 |
| librustc_privacy |   1.2 |
| librand |   1.2 |
| libunicode |   0.7 |
| rustbook |   0.6 |
| librustc_llvm |   0.5 |
| liballoc |   0.4 |
| libfmt_macros |   0.2 |
| libgraphviz |   0.1 |
| liblog |   0.1 |
| libarena |   0.1 |
| libflate |   0.0 |
| librustc_bitflags |   0.0 |
| liblibc |   0.0 |


(Collected with [this diff](https://gist.github.com/huonw/fde4691d8ac597a32a0f) + some emacs macros etc.)

File: src/librustc/lint/builtin.rs [hunk 1]
```diff
@@ -1789,6 +1791,194 @@ impl LintPass for Stability {
...
+                let target_idx = edge.target();
+                let target_cfg_id = target_idx.node_id();
+                if !visited.contains(&target_cfg_id) {
+                    work_queue.push(target_idx)
+                }
+                true
+            });
+        }
+
+        // check the number of sell calls because a function that
```
Comment (eddyb) [inline src/librustc/lint/builtin.rs]: Should "sell" be "self"?

**Pull request merged**